[OptimistLabyrinth] 🚀 2단계 - 문자열 덧셈 계산기

OptimistLabyrinth opened the pull request:
아직 완료는 아니지만 제대로 하고 있는지 피드백을 받아보는 게 좋을 것 같아서
중간과정에서 PR 생성했습니다.

혹시 제가 잘못된 방향으로 가고 있다면 어떤식으로 수정해야할지
보다 중점적으로 신경써야 하는 부분은 어디인지
말씀해주시면 감사하겠습니다.

Commit:
feat: [2단계] 문자열 덧셈 계산기 시작

Commit:
test: RED - 입력이 null 또는 ""(empty string) 인 경우

여기서는 RED 상태
입력 문자열이 null 또는 empty string 인 경우에 대한 유닛 테스트 코드 작성

Commit:
feat: GREEN - zeroIfNullOrEmpty 테스트 통과하게 만들기

StringCalculator 클래스에 생성자, 메서드를 적절하게 추가해서
테스트코드가 통과하도록 만든다

Commit:
feat: REFACTOR - calculate 메서드 null, empty string 검사 로직 구현

org.apache.commons:commons-lang3 패키지 설치
기존에 사용하고 잇던 패키지 최신 버전으로 업데이트

StringCalculator 클래스 calculate 메서드에서 StringUtils.isEmpty 사용

Commit:
test: 테스트 케이스에 대해 DisplayName 한글 설명 추가

Commit:
feat: 구분자를 찾아내는 DelimiterFinder 에 대한 TDD 방식 구현 준비

Commit:
test: RED - 입력에 커스텀 구분자를 표시하는 문자가 있는지 확인하는 테스트 코드

여기까지는 RED 상태
커스텀 구분자를 감싸는 이중 슬래시, 뉴라인 캐릭터가 모두 들어있는지 확인하는
메서드에 대한 테스트 코드 작성

Commit:
feat: GREEN-REFACTOR - 메서드 ifIncludeBothSlashAndNewline 구현

이제는 GREEN 상태
이중 슬래시(`//`)와 뉴라인(`\n`) 을 모두 포함하는 경우 true 를 반환한다

Commit:
chore: 오타 수정

Commit:
test: RED - 이중 슬래시, 뉴라인 문자 사이에 있는 커스텀 구분자 유효성 확인하는 메서드의 테스트 코드

여기까지는 RED 상태
이중 슬래시, 뉴라인 사이에 유효한 문자가 있는지 확인하는
메서드에 대한 테스트 코드 작성

Commit:
feat: GREEN - customDelimiterExistsInBetween 메서드 테스트 통과는 했지만 너무 복잡해서 분리 필요

이제는 GREEN 상태
그러나 customDelimiterExistsInBetween 메서드 내부 로직이 너무 복잡해서
오류 발생할 여지가 너무 많다
메서드를 분리해야 한다

Commit:
test: RED - 이중 슬래시, 뉴라인이 서로 붙어있는 확인하는 메서드 준비

이중 슬래시, 뉴라인이 서로 붙어있는지 확인하는 메서드에 대한
테스트 코드 작성

Commit:
feat: 메서드 분리는 private 메서드로 하기로 결정

새롭게 만든 메서드
- noSpaceBetweenDoubleSlashAndNewline
- customDelimiterMoreThanOne
- isCustomDelimiterDigit

Commit:
test: RED - 입력에서 구분자 찾는 메서드 테스트 코드 작성

입력 문자열에서 구분자를 찾는 메서드 find 에 대한 테스트 코드 작성
규칙에 따라서 판단 후에 올바르지 않다면 기본값 구분자를 반환하고
커스텀 구분자를 사용한다면 이 커스텀 구분자를 전달한다

Commit:
feat: GREEN-REFACTOR - find 메서드 구현

Commit:
feat: DelimiterFinder 사용해서 확인한 구분자 기준으로 입력값 split

StringCalculator 클래스에 splitInput 메서드 추가

Commit:
fix: 기본 구분자는 콜론(:) 이다

오류 수정
세미콜론(;) 은 기본 구분자가 아니다

Commit:
feat: 입력값을 구분자로 split 하는 메서드 구현

String 클래스의 split 메서드는 아규먼트(argument) 로
정규표현식 문자열을 받는다
정규표현식에서 사용하는 예약 문자가 커스텀 구분자가 되는 경우에 한해서 특별한 처리가 필요하고
ifDelimiterRegexReservedWord 에서 확인한 후
splitInput 메서드에서

```java
String.format("\\%s", delimiter);
```

와 같이 처리한다

Commit:
feat: InputSplitter 클래스로 로직 분리하기

StringCalculator 클래스에서 splitInput 메서드에 해당하는 로직이
꽤나 복잡해져서
InputSplitter 클래스로 분리하기

테스트 코드도 StringCalculatorTest 에서 InputSplitterTest 로 이동

Commit:
feat: split 이후 개별 토큰을 저장하는 Operand 클래스 준비

seokhongkim commented on build.gradle:
```diff
@@ -12,9 +12,10 @@ repositories {
-    testImplementation 'org.assertj:assertj-core:3.19.0'
-    testImplementation 'org.junit.jupiter:junit-jupiter:5.7.1'
-    testImplementation 'org.mockito:mockito-inline:3.6.0'
+    implementation 'org.apache.commons:commons-lang3:3.12.0'
```
미션 진행은 외부 라이브러리 의존성 추가 없이 해보시면 좋겠습니다. :)
실무에서 apache commons 에 있는 라이브러리에서 제공해주는 API가 유용하고 많이 사용하지만, 학습을 위해 미션에서는 직접 구현해보시면 더 도움이 될 것 같습니다.

라이브러리 버전 변경도 버전에 따라 추가되는 부분이 있을 수 있지만, 미션 설계 의도에 따른 진행을 위해 기존 버전으로 해보시면 좋을 것 같습니다.
혹시 이 부분에 대해 강사분과 따로 논의를 마치셨다면, 말씀 부탁드립니다.

OptimistLabyrinth replied:
아하 그렇군요 ㅎㅎ 한번 시도해보겠습니다.

seokhongkim commented on src/main/java/string/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,24 @@
+
+    public int calculate() {
+        if (isEmpty(input)) {
+            return 0;
```
기본으로 반환하는 값임을 명시적인 이름을 부여한 상수로 관리해보시면 좋겠습니다.

seokhongkim commented on src/main/java/string/calculator/DelimiterFinder.java:
```diff
@@ -0,0 +1,51 @@
+        if (ifIncludeBothSlashAndNewline() && customDelimiterExistsInBetween()) {
+            return ultimateCustomDelimiter;
+        }
+        return "[:,]";
```
default delimiter 임을 명시적인 이름이 있는 상수로 관리해보시면 좋겠습니다.

seokhongkim commented on src/main/java/string/calculator/DelimiterFinder.java:
```diff
@@ -0,0 +1,51 @@
+        return input.contains("//") && input.contains("\\n");
+    }
+
+    public boolean customDelimiterExistsInBetween() {
```
custom delimiter 를 찾는 부분을 미션 안내에 있는 regular expression 을 사용해서 구현해보시면 좋겠습니다.
> https://edu.nextstep.camp/s/urD8knTu/ls/sDuxkHkp

OptimistLabyrinth replied:
ㅎㅎ 그렇군요
좀 더 자세히 읽어보겠습니다.

seokhongkim commented on src/main/java/string/calculator/InputSplitter.java:
```diff
@@ -0,0 +1,40 @@
+package string.calculator;
+
+public class InputSplitter {
```
`InputSplitter` 의 로직도 너무 복잡하게 생각하지 않으셔도 괜찮을 것 같습니다.
> https://edu.nextstep.camp/s/urD8knTu/ls/sDuxkHkp

미션 설명에 있는 힌트를 참고하시어 최대한 단순하게 구현해보시면 좋겠습니다.

OptimistLabyrinth replied:
하... 다음 단계부터는 과제 안내문을 잘 읽고 해야겠네요

ㅋㅋㅋㅋㅋ 이렇게 짧게 할 수 있는 작업을 어제 몇 시간을 낭비한건지

seokhongkim reviewed (changes requested):
기성님 안녕하세요.
로또 미션을 함께 할 리뷰어 김석홍입니다. 만나서 반갑습니다! 😀 

1단계 미션 리뷰 요청 주셔서 코멘트를 우선 남겼습니다.

이번 단계의 다음 리뷰 요청을 포함하여 리뷰 요청은 기능 요구사항을 구현하신 상태에서 주시면 좋겠습니다.
미션 구현 방법에 정답은 없고, 구현 중간 단계와 구현이 완료된 후의 코드나 설계가 달라질 수 있다고 생각합니다. 직접 끝까지 고민해서 구현해보신 후 리뷰와 토론을 통해 더 깊은 고민을 할 수 있다고 생각합니다. :)

구현하는 과정에서 너무 어렵게 느껴지시는 부분이 있거나, 고민되는 점이 있다면 슬랙 DM으로 말씀해주시면 같이 이야기 나눠볼 수 있을 것 같습니다!

아직 구현이 완료되지 않은 상황이라 코멘트는 조금만 남겼습니다.
코멘트 참고하시어 기능 구현 완료 후 다시 리뷰 요청 부탁드립니다. 🙏

Commit:
refactor: third-party 라이브러리 되도록 사용하지 않기

apache.commons 라이브러리는 학습에 도움이 되지 않으므로
사용하지 않는다

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007879000](https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007879000)

Commit:
refactor: 덧셈 계산기에서 기본반환값을 이름 있는 상수로 관리하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007880392](https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007880392)

Commit:
refactor: defatult delimiter 를 이름 있는 상수로 관리하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007884045](https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007884045)

Commit:
refactor: regex 사용해서 구분자 찾기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007886346](https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007886346)

Commit:
refactor: 미션 설명을 참고해서 InputSplitter 클래스 로직 간소화

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007887143](https://github.com/next-step/java-lotto-pro/pull/756#discussion_r1007887143)

Commit:
test: (RED) Operand 생성자에 대한 실패하는 테스트 코드 만들기

Commit:
feat: (GREEN) Operand 생성자에 대한 테스트 코드 성공하도록 만들기

Operand 클래스는 private int 필드를 하나 가지고 있고
이 필드는 생성자에서 초기화한다

Commit:
feat: (REFACTOR-GREEN) 음수값을 입력하면 예외가 발생한다

Commit:
feat: (REFACTOR-GREEN) Operand 생성자 파라미터 타입을 String 변경

Commit:
test: (REFACTOR-GREEN) 숫자가 아닌 값을 이용해서 Operand 객체 생성하려고 하면 예외 발생

Commit:
feat: StringCalculator 클래스 convertToOperand 메서드 만들기

입력 문자열을 split 한 결과를
Operand 타입의 배열로 변환한다

Commit:
feat: (REFACTOR-GREEN) Operand 클래스에 add 메서드 추가, equals, hashCode 메서드 오버라이드

add 메서드에 대한 테스트 코드도 추가

Commit:
feat: SumOfOperands 클래스 준비

Commit:
feat: (RED-GREEN-REFACTOR) SumOfOperands 클래스 만들기 및 구현

SumOfOperands 생성자에서 파라미터로 Operand[] operands 를 받고
sum 결과값을 인스턴스 변수로 저장한다

SumOfOperands 클래스에 대한 단위 테스트 코드 작성

Commit:
feat: Operand[] operands 배열의 총합(sum) 결과값 반환하도록 구현

StringCalculator 클래스의 calculate 메서드에서
최종 덧셈 결과값을 반환하도록 구현한다

Commit:
chore: 클래스 이름 변경

StringCalculator -> StringAddCalculator

Commit:
test: 숫자 하나만 입력하는 경우에 대한 테스트 케이스 추가

Commit:
test: InputSplitter 단위 테스트에 입력이 숫자 하나인 경우도 추가

OptimistLabyrinth commented:
@seokhongkim 

2단계 "덧셈 계산기" 전체 구현 완료해서 리뷰 요청 드렸습니다
시간 되실 때 살펴보고 피드백 부탁드리겠습니다 :)

seokhongkim commented on src/main/java/string/calculator/InputSplitter.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.regex.Pattern;
+
+public class InputSplitter {
+    private static final String DEFAULT_DELIMITER_REGEX = "[:,]";
```
상수로 관리하신 것 좋습니다. 👍

seokhongkim commented on src/main/java/string/calculator/InputSplitter.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public String[] split() {
+        final Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(input);
```
`Pattern.compile("//(.)\n(.*)")` 의 결과물인 `Pattern`도 thread sfae 하니, 미리 하나를 만들어서 상수로 관리해보셔도 좋을 것 같습니다.
이게 무엇을 의미하는지 적절한 이름도 추가해보실 수 있을 것 같아요!

seokhongkim commented on src/main/java/string/calculator/Operand.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Objects;
+
+public class Operand {
```
값 객체로 관리하시고 필요한 밸리데이션을 수행하신 것 좋습니다. 👍 
`add` 기능에서도 불변으로 유지하신 것도 좋습니다. :)

seokhongkim commented on src/main/java/string/calculator/Operand.java:
```diff
@@ -0,0 +1,43 @@
+
+    public Operand add(Operand other) {
+        int addResult = value + other.value;
+        return new Operand(Integer.toString(addResult));
```
문자열로 변경 후 다시 정수형으로 변경해야 한다면, `Operand` 에서 정수형 값을 받는 생성자를 오버로딩해서 사용해보셔도 좋을 것 같네요!

OptimistLabyrinth replied:
아아 그 생각을 못했네요. 제안 고맙습니다 👍👍

seokhongkim commented on src/test/java/string/calculator/InputSplitterTest.java:
```diff
@@ -0,0 +1,36 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class InputSplitterTest {
```
클래스별 역할을 잘 분리해주셨고, 단위 테스트도 꼼꼼하게 잘 작성해주셨네요! 💯 

`StringAddCalculatorTest` 에서 정상 입력인 경우 더한 값을 잘 계산해주는지에 대한 테스트도 추가되면 더 좋을 것 같습니다. :)

seokhongkim reviewed (approved):
기성님, 2단계 미션 잘 마무리 해주셨습니다. 👍 
다음 단계 미션 진행을 위해 2단계 PR은 이만 머지하도록 하겠습니다.

피드백을 조금 남겼는데, 3단계 미션 진행하실 때 참고해보셔도 괜찮을 것 같습니다.
3단계 미션에서 뵙겠습니다! 😀

OptimistLabyrinth commented:
정성스럽게 봐주셔서 감사합니다 😄😄
3단계 미션 진행할때 피드백 내용 참고해서 해보겠습니다 🙌
